### PR TITLE
Global slice and sort (#397)

### PR DESCRIPTION
**Pull request by nickhand**
Adds global sort and global slice functions to CatalogSource (and CatalogMesh) objects

### CODE REVIEW

**Comment by nickhand** [comment]
Is this what you had in mind @rainwoodman ?

File: nbodykit/base/catalog.py [hunk 1]
```diff
@@ -1165,6 +1165,117 @@ def csize(self):
             return NotImplemented
         return self._csize
 
+    def gslice(self, start, stop, end=1):
+        """
+        Execute a global slice of a CatalogSource.
+
+        .. note::
+            After the global slice is performed, the data is scattered
```
Comment (rainwoodman) [inline nbodykit/base/catalog.py]: I wonder if it is useful to avoid the scattering with an option?

**Comment by rainwoodman** [comment]
I have to find a larger chunk of time to read through this. But briefly reading this, it seems not supporting slicing a ColumnAccessor. Could it be useful to globally slice a columnaccessor?

File: nbodykit/base/catalog.py (hunk 1 above)
Comment (nickhand) [reply nbodykit/base/catalog.py]: I am not sure I can think of a use case now, but it doesn't hurt to include the option, I think

**Comment by nickhand** [comment]
No problem, let's push this to v0.2.8 anyways. 

I think I agree. It would be nice to call both sort() and gslice() on a ColumnAccessor, I think

File: nbodykit/base/catalog.py (hunk 1 above)
Comment (rainwoodman) [reply nbodykit/base/catalog.py]: Avoiding the scattering probably means the compute can be deferred further?

**Comment by rainwoodman** [comment]
The intended main use case of gslice is for abundance matching. Gslicing on a column accessor does not fit into this context.

Currently a gslice will trigger a full scan of the catalog, does it?

File: nbodykit/base/catalog.py (hunk 1 above)
Comment (nickhand) [reply nbodykit/base/catalog.py]: yes I think so. I wonder if it is possible to implement this with no compute() calls?  We can't use Scatter/GatherArray for dask arrays but we can bcast the appropriate slices of them?

**Comment by nickhand** [comment]
As it's written now, it computes the whole catalog and re-scatters it evenly after the slice. In theory, we could pass around the dask arrays instead?

**Comment by rainwoodman** [comment]
Perhaps leave a few comments in the code to describe how this 'could be done' without actually doing it. Because in reality if we are abundance matching we probably can easily read in the entire data set in most cases (especially we already done a sorting) The IO is cut off from that point when we have the abundance matched catalog, and it's not a bad thing.

The situation may change if we are goaling to make catalogs for LSST or BOA.

**Comment by nickhand** [comment]
This looks ready too so I'll merge this in for 0.2.8

**Pull request merged**